clients/src/main/java/org/apache/kafka/common/security/scram/ScramMessages.java

```diff
@@ -0,0 +1,272 @@
+     */
+    public static class ClientFirstMessage extends AbstractScramMessage {
+        private static final Pattern PATTERN = Pattern.compile(String.format(
+                "n,(a=(?<authzid>%s))?,%sn=(?<saslname>%s),r=(?<nonce>%s)%s",
+                SASLNAME,
+                RESERVED,
+                SASLNAME,
+                PRINTABLE,
+                EXTENSIONS));
+
+        private final String saslName;
+        private final String nonce;
+        private final String authorizationId;
+        public ClientFirstMessage(byte[] messageBytes) throws SaslException {
+            String message = toMessage(messageBytes);
+            Matcher matcher = PATTERN.matcher(message);
+            if (!matcher.matches())
+                throw new SaslException("Invalid SCRAM client first message format: " + message);
+            String authzid = matcher.group("authzid");
+            this.authorizationId = authzid != null ? authzid : "";
+            this.saslName = matcher.group("saslname");
+            this.nonce = matcher.group("nonce");
+        }
+        public ClientFirstMessage(String saslName, String nonce) {
+            this.saslName = saslName;
+            this.nonce = nonce;
+            this.authorizationId = ""; // Optional authzid not specified in gs2-header
+        }
+        public String saslName() {
+            return saslName;
+        }
+        public String nonce() {
+            return nonce;
+        }
+        public String authorizationId() {
+            return authorizationId;
+        }
+        public String gs2Header() {
+            return "n," + authorizationId + ",";
+        }
+        public String clientFirstMessageBare() {
+            return String.format("n=%s,r=%s", saslName, nonce);
+        }
+        String toMessage() {
+            return gs2Header() + clientFirstMessageBare();
+        }
+    }
+
+    /**
+     * Format:
+     *     [reserved-mext ","] nonce "," salt "," iteration-count ["," extensions]
+     * Limitations:
+     *     Extensions are ignored.
+     *
+     */
+    public static class ServerFirstMessage extends AbstractScramMessage {
+        private static final Pattern PATTERN = Pattern.compile(String.format(
+                "%sr=(?<nonce>%s),s=(?<salt>%s),i=(?<iterations>[0-9]+)%s",
+                RESERVED,
+                PRINTABLE,
+                BASE64,
+                EXTENSIONS));
+
+        private final String nonce;
+        private final byte[] salt;
+        private final int iterations;
+        public ServerFirstMessage(byte[] messageBytes) throws SaslException {
+            String message = toMessage(messageBytes);
+            Matcher matcher = PATTERN.matcher(message);
+            if (!matcher.matches())
+                throw new SaslException("Invalid SCRAM server first message format: " + message);
+            try {
+                this.iterations = Integer.parseInt(matcher.group("iterations"));
+                if (this.iterations <= 0)
```

Comment:
It seems that checks like this is only done when the config change is propagated to the broker? Ideally, we want to fail the ConfigChangeCommand if the input is not valid.